bootstrap/src/main/java/io/airlift/bootstrap/Bootstrap.java

```diff
@@ -322,17 +322,24 @@ private static ColumnPrinter makePrinterForConfiguration(ConfigurationFactory co
         Map<String, String> replaced = new HashMap<>();
         properties.forEach((propertyKey, propertyValue) -> {
             Matcher matcher = ENV_PATTERN.matcher(propertyValue);
-            if (!matcher.matches()) {
+            List<String> envNames = new ArrayList<>();
+            while (matcher.find()) {
+                envNames.add(matcher.group(1));
```

Comment:
For finding multiple matches and replacing them all it would more idiomatic to use `java.util.regex.Matcher#appendReplacement` and `java.util.regex.Matcher#appendTail` methods.

It yields for simpler code and also prevents "recursive" processing inconsistencies

The final code could similar to

```java
StringBuffer replacedPropertyValue = new StringBuffer();
Matcher matcher = ENV_PATTERN.matcher(propertyValue);
while (matcher.find()) {
    String envName = matcher.group(1);
    String envValue = environment.get(envName);
    if (envValue == null) {
        onError.accept(propertyKey, format("Configuration property '%s' references unset environment variable '%s'", propertyKey, envName));
        return;
    }
    matcher.appendReplacement(replacedPropertyValue, quoteReplacement(envValue));
}
matcher.appendTail(replacedPropertyValue);
replaced.put(propertyKey, replacedPropertyValue.toString());
```

Reply:
@electrum the code i suggested above fails at modernizer
```
[INFO] --- modernizer-maven-plugin:1.7.1:modernizer (modernizer) @ bootstrap ---
[ERROR] /Users/xxx/repos/airlift/bootstrap/src/main/java/io/airlift/bootstrap/Bootstrap.java:325: Prefer java.lang.StringBuilder
```

thoughts?

Reply:
@findepi Thank you for the review. I will fix the "recursive replacement" problem as suggested.

I think the modernizer violation is caused by using appendReplacement with StringBuffer?
Using StringBuilder instead of StringBuffer can resolve the error but it requires language level 9 and StringBuilder is not thread-safe.

Should we exclude this violation? I am not familiar with the modernizer plugin. I have tried using @SuppressModernizer annotation and adding exclusion configuration in boostrap/pom.xml but still get the same error when running the modernizer plugin.

Reply:
@findepi I realized the problem is not caused by using appendReplacement with StringBuffer.
Instantiating a StringBuffer causes the error while only declaring a StringBuffer does not.

Reply:
Using `StringBuilder` instead of `StringBuffer` can resolve the error but it requires language level 9 to use `appendReplacement` and `appendTail`.

Should we exclude this violation?
The error is eliminated after adding the following in `bootstrap/pom.xml`.
```
<build>
    <plugins>
        <plugin>
            <groupId>org.gaul</groupId>
            <artifactId>modernizer-maven-plugin</artifactId>
            <configuration>
                <exclusions>
                    <exclusion>java/lang/StringBuffer."&lt;init&gt;":()V</exclusion>
                </exclusions>
            </configuration>
        </plugin>
    </plugins>
</build>
```

Reply:
Let's exclude it for now, with TODO comment to remove after updating to Java 9.